In [7]:
import pandas as pd
import numpy as np 
from typing import List
from typing import Tuple
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression
from typing import List
from typing import Tuple
df = pd.read_csv('Credit_Card.csv')
label_df = pd.read_csv('Credit_card_label.csv')
merged_df = pd.merge(df, label_df, on='Ind_ID')

def clean_and_encode_data(merged_df: pd.DataFrame, 
                          numerical_cols: List[str] = ['Annual_income', 'Birthday_count', 'Employed_days'], 
                          categorical_cols: List[str] = ['GENDER', 'Car_Owner', 'Propert_Owner', 'Type_Income', 
                                                         'EDUCATION', 'Marital_status', 'Housing_type', 'Type_Occupation'], 
                          occupation_col: str = 'Type_Occupation') -> pd.DataFrame:
    for col in numerical_cols:
        df[col] = df[col].fillna(df[col].median())
    
    merged_df.dropna(subset=[occupation_col], inplace=True)
    
    merged_df['GENDER'] = merged_df['GENDER'].fillna(merged_df['GENDER'].mode()[0])
    
    df_encoded = pd.get_dummies(merged_df, columns=categorical_cols, drop_first=True)

    df_encoded = df_encoded.replace([np.inf, -np.inf], np.nan).fillna(0)

    df_encoded = df_encoded.astype(int)

    sc = StandardScaler()
    df_encoded[numerical_cols] = sc.fit_transform(df_encoded[numerical_cols])
    return df_encoded
df_encoded = clean_and_encode_data(merged_df)
df_encoded

Ind_ID  CHILDREN  Annual_income  Birthday_count  Employed_days  \
8     5010864         1       2.047160       -0.909858       0.842207   
9     5010868         1       2.047160       -0.909858       0.842207   
10    5010869         1       2.047160       -0.909858       0.842207   
11    5018498         0      -0.878721       -1.110597       0.706227   
12    5018501         0      -1.610191       -1.110597       0.706227   
...       ...       ...            ...             ...            ...   
1542  5118268         1       1.315690        0.867338      -0.357267   
1543  5028645         0      -1.610191        0.696052       0.210993   
1544  5023655         0       0.218485        1.142482       0.619351   
1545  5115992         2      -0.147250        0.381638       0.087185   
1546  5118219         0       0.584220       -0.165549       0.856056   

      Mobile_phone  Work_Phone  Phone  EMAIL_ID  Family_Members  ...  \
8                1           0      1         1               3  ...   
9                1           0      1         1               3  ...   
10               1           0      1         1               1  ...   
11               1           1      1         0               2  ...   
12               1           1      1         0               2  ...   
...            ...         ...    ...       ...             ...  ...   
1542             1           0      1         0               3  ...   
1543             1           0      0         0               2  ...   
1544             1           0      0         0               1  ...   
1545             1           0      0         0               4  ...   
1546             1           1      1         0               2  ...   

      Type_Occupation_Laborers  Type_Occupation_Low-skill Laborers  \
8                            0                                   0   
9                            0                                   0   
10                           0                                   0   
11                           0                                   0   
12                           0                                   0   
...                        ...                                 ...   
1542                         0                                   0   
1543                         0                                   0   
1544                         0                                   0   
1545                         0                                   0   
1546                         0                                   0   

      Type_Occupation_Managers  Type_Occupation_Medicine staff  \
8                            0                               0   
9                            0                               0   
10                           0                               0   
11                           0                               0   
12                           0                               0   
...                        ...                             ...   
1542                         0                               0   
1543                         1                               0   
1544                         0                               0   
1545                         1                               0   
1546                         0                               0   

      Type_Occupation_Private service staff  Type_Occupation_Realty agents  \
8                                         0                              0   
9                                         0                              0   
10                                        0                              0   
11                                        0                              0   
12                                        0                              0   
...                                     ...                            ...   
1542                                      0                              0   
1543   

In [16]:

X = df_encoded.drop(columns=['label', 'Ind_ID'])
y = df_encoded['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

log_reg = LogisticRegression(random_state=42)
log_reg.fit(X_train, y_train)

y_pred = log_reg.predict(X_test)

log_reg.score(X_test, y_test)


0.8584905660377359

In [12]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [14]:
lda = LinearDiscriminantAnalysis(n_components=1)
X_train_lda = lda.fit_transform(X_train, y_train)
X_test_lda = lda.transform(X_test)
model = LogisticRegression(random_state=42).fit(X_train_lda, y_train)

model.score(X_test_lda, y_test)

0.8537735849056604

The slight drop in my models score from 0.858 to 0.853. It  suggests that dimensionality reduction may have removed some useful information, causing a minor loss in predictive power. But since the drop isn't that big it means that there wasn't too much useful information removed